src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs

```diff
@@ -1,45 +1,273 @@
+        _connectionOptionsCallback = options.ConnectionOptionsCallback;
+        _acceptQueue = Channel.CreateBounded<PendingConnection>(new BoundedChannelOptions(options.ListenBacklog) { SingleWriter = true });
+
+        // Start the listener, from now on MsQuic events will come.
+        using MsQuicBuffers alpnBuffers = new MsQuicBuffers();
+        alpnBuffers.Initialize(options.ApplicationProtocols, applicationProtocol => applicationProtocol.Protocol);
+        QuicAddr address = options.ListenEndPoint.ToQuicAddr();
+        if (options.ListenEndPoint.Address.Equals(IPAddress.IPv6Any))
+        {
+            // For IPv6Any, MsQuic would listen only for IPv6 connections. This would make it impossible
+            // to connect the listener by using the IPv4 address (which could have been e.g. resolved by DNS).
+            // Using the Unspecified family makes MsQuic handle connections from all IP addresses.
+            address.Family = QUIC_ADDRESS_FAMILY_UNSPEC;
+        }
+        ThrowIfFailure(MsQuicApi.Api.ApiTable->ListenerStart(
+            _handle.QuicHandle,
+            alpnBuffers.Buffers,
+            (uint)alpnBuffers.Count,
+            &address),
+            "ListenerStart failed");
+
+        // Get the actual listening endpoint.
+        LocalEndPoint = MsQuicParameterHelpers.GetIPEndPointParam(MsQuicApi.Api, _handle, QUIC_PARAM_LISTENER_LOCAL_ADDRESS, options.ListenEndPoint.AddressFamily);
+    }
+
+    public async ValueTask<QuicConnection> AcceptConnectionAsync(CancellationToken cancellationToken = default)
+    {
+        ThrowIfDisposed();
+
+        try
+        {
+            while (true)
+            {
+                PendingConnection pendingConnection = await _acceptQueue.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+                await using (pendingConnection.ConfigureAwait(false))
+                {
+                    var connection = await pendingConnection.FinishHandshakeAsync(cancellationToken).ConfigureAwait(false);
+                    // Handshake failed, discard this connection and try to get another from the queue.
+                    if (connection is null)
+                    {
+                        continue;
+                    }
+
+                    return connection;
+                }
+            }
+        }
+        catch (ChannelClosedException ex) when (ex.InnerException is not null)
+        {
+            ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
+            throw;
+        }
+    }
 
-        public static ValueTask<QuicListener> ListenAsync(QuicListenerOptions options, CancellationToken cancellationToken = default)
+    private unsafe int HandleListenerEvent(ref QUIC_LISTENER_EVENT listenerEvent)
+    {
+        if (listenerEvent.Type == QUIC_LISTENER_EVENT_TYPE.NEW_CONNECTION)
         {
-            if (!IsSupported)
+            ref var data = ref listenerEvent.NEW_CONNECTION;
```

Comment:
Concrete type instead of `var`.

Reply:
This is the only place where it's not reasonable since the type name is generated from an anonymous union and looks like this: `_Anonymous_e__Union._NEW_CONNECTION_e__Struct`. So I'd prefer to keep the `var` here.